### PR TITLE
CCv0: SEV prelaunch attestation

### PR DESCRIPTION
**Pull request by jimcadden**
**TODO:**
- [x] Include CCv0 components in guest 
- [x] Validate secret are received by guest 
- [x] Enable measurement validation in GOP
- [x] Rebase onto CCv0 branch
- [x] clean up debug prints 
- [x] Remove hard-coded sev-policy
- [x] Add `kernel-hashes=on` for QEMU 6.2+

### CODE REVIEW

File: src/runtime/virtcontainers/qemu_amd64.go
Comment (Jakob-Naucke) [inline src/runtime/virtcontainers/qemu_amd64.go:292]: minor nit just from glancing over this: do you want to make `SEV attestation` a log field rather than explicitly printing it every time?

Comment (jodh-intel) [reply src/runtime/virtcontainers/qemu_amd64.go:292]: Agreed - https://github.com/kata-containers/tests/tree/main/cmd/log-parser.

File: src/runtime/virtcontainers/qemu_amd64.go
Comment (jodh-intel) [inline src/runtime/virtcontainers/qemu_amd64.go:356]: Please minimise log calls - these all relate to the same thing so can be a single call with log fields: https://github.com/kata-containers/tests/tree/main/cmd/log-parser.

File: src/runtime/virtcontainers/qemu_amd64.go
Comment (jodh-intel) [inline src/runtime/virtcontainers/qemu_amd64.go:360]: - I thought reflect had a relatively significant performance impact. Maybe that's no longer the case but do we really *have* to use this package?
- What if `qemu.Object` changes name?

**Review by jodh-intel — COMMENTED** [review]
Thanks @jimcadden.

First pass comments.

**Comment by jodh-intel** [comment]
@jimcadden - any updates on this?

**Comment by jodh-intel** [comment]
Hi @jimcadden - are you still working on this or should we close for now?

**Comment by Jakob-Naucke** [comment]
@jodh-intel (it's Thanksgiving but) I think this is required for SEV

**Comment by sameo** [comment]
> Hi @jimcadden - are you still working on this or should we close for now?

Let's keep it open for now, I assume Jim is going to send an updated version.

File: src/runtime/virtcontainers/qemu_amd64.go
Comment (jimcadden) [reply src/runtime/virtcontainers/qemu_amd64.go:381]: Thanks @jodh-intel, libcontainer is a better approach here. Ideally, we envision making the gRPC calls to the guest-owner's attestation service from within the kata shim, avoiding this external script all together. So, for now, this external script (and the prepended namespace hop w/ `nsenter`) is a quick hack for the demo.

File: src/runtime/virtcontainers/qemu_amd64.go
Comment (jimcadden) [reply src/runtime/virtcontainers/qemu_amd64.go:360]: Agreed - this is very ugly. But, I'm not really sure another way to go about it. What is happening here is that we  traverse a list of generic-typed Device objects to get a reference to a previously-created `sev-guest` QEMU object. We assume there's only one `qemu.Object` on the list, but that's not necessarily true. 

> I thought reflect had a relatively significant performance impact.
Not sure. We traverse the device list and call reflect twice during the VM setup and launch. The attestation process is already pretty expensive and the device list is very short, O(1), so I haven't noticed a performance hit. 

> What if qemu.Object changes name? 
And what if there's more than one `qemu.Object`? This piece of code is full of dangerous assumptions. 

Any thoughts @Jakob-Naucke ?

File: src/runtime/virtcontainers/qemu_amd64.go
Comment (jimcadden) [reply src/runtime/virtcontainers/qemu_amd64.go:292]: Yes, I'll clean up these debug prints

File: src/runtime/virtcontainers/qemu_amd64.go
Comment (jimcadden) [reply src/runtime/virtcontainers/qemu_amd64.go:47]: @jodh-intel Yes, will do

@fitzthum What do you think? Call `gop-*.py` a "quick hack" and plan to rewrite in go? This is beyond scope of CCv0

Comment (fitzthum) [reply src/runtime/virtcontainers/qemu_amd64.go:47]: Ideally it would all be part of the shim. `gop-client.py` isn't a particularly sophisticated piece of code. We might lose some modularity by adding the functionality into the shim directly, though. It would set the protocol for any SEV(-ES) KBS.

**Comment by jimcadden** [comment]
Hi @jodh-intel @sameo, Apologies for the radio silence. I have bandwidth again and my aim is to get this PR cleaned up over the holiday break :)

File: src/runtime/virtcontainers/qemu_amd64.go
Comment (Jakob-Naucke) [reply src/runtime/virtcontainers/qemu_amd64.go:360]: > Any thoughts @Jakob-Naucke ?

I wouldn't know of another way unless we introduce something annotation-like to govmm (which would be too many LOC to justify it IMO).

> what if there's more than one `qemu.Object`?

Aren't you still checking all elements individually?

Also, I've been trying to play with this, and the signature of `ExecuteSEVInjectLaunchSecret` doesn't match confidential-containers-demo/govmm/sev-attestation. Does this need an update?

Comment (jimcadden) [reply src/runtime/virtcontainers/qemu_amd64.go:360]: >  Does this need an update?

Yup, nice catch! All synced

**Comment by jodh-intel** [comment]
@jimcadden - 10-4! :smile:

File: src/runtime/virtcontainers/sandbox.go
Comment (jimcadden) [reply src/runtime/virtcontainers/sandbox.go]: removed this work-around. No longer needed

**Comment by jodh-intel** [comment]
Any update on this @jimcadden?

File: src/runtime/cli/config/configuration-qemu.toml.in
Comment (jimcadden) [reply src/runtime/cli/config/configuration-qemu.toml.in]: Skipped. QEMU restriction should rather be around the `confidential_guest` argument and/or the description of Confidential Containers as a whole.

File: src/runtime/virtcontainers/qemu_amd64.go
Comment (jimcadden) [reply src/runtime/virtcontainers/qemu_amd64.go:345]: I looked into this and found that a `libcontainer` solution would be far more complex. Calling out to `docker` may be a  improvement here, but that adds more external dependencies to the runtime. Unless there's major protest, I say we keep using `exec.Command` and `nsenter` for CCv0, blemishes and all.

Comment (jimcadden) [reply src/runtime/virtcontainers/qemu_amd64.go:345]: @fitzthum and I are in discussion about an improved KBS for SEV/-ES, most likely to be written and rust and deployed in a container, which would replace these`sevGuestOwnerProxyClient` calls.

**Comment by jimcadden** [comment]
I think this is ready for a re-review. I intent to add some addition instructions about the GOP to the commit message. Otherwise I am happy with the changes.

**Review by jodh-intel — CHANGES_REQUESTED** [review]
Thanks @jimcadden, but the `vendor/` directory must not be modified directly. Please see:

- https://github.com/kata-containers/community/blob/main/CONTRIBUTING.md#re-vendor-prs

**Comment by jimcadden** [comment]
> Thanks @jimcadden, but the `vendor/` directory must not be modified directly. Please see:

Yes, I know. Commit #f72a6d5 is a convenience intended only for the CCv0 branch, [which already contains changes to the `vendor/` directory](https://github.com/kata-containers/kata-containers/commit/822efa344d8eb1846bf638703dafdc7698693b76 ). I'm happy to remove it if need be.

**Comment by jodh-intel** [comment]
Oh - What's the plan for resolving those changes to `vendor/` and removing the technical debt then? That situation could get out of control pretty quickly I'd imagine.

/cc @fidencio

**Comment by jimcadden** [comment]
> Oh - What's the plan for resolving those changes to `vendor/` and removing the technical debt then? That situation could get out of control pretty quickly I'd imagine.
> 
> /cc @fidencio

My understanding is that the CCv0 branch is not intended for direct upstream. Instead, individual components within the CCv0 mvp will be split off into new PRs->main.

**Comment by jimcadden** [comment]
Also, I will remove commit [3025](https://github.com/kata-containers/kata-containers/pull/3025/commits/f682220d596c99845082298acbea3501375a6688) once #3223 lands

**Comment by fidencio** [comment]
A head's up that may affect your PR.

In the Architecture Committee meeting from January 25th, 2022, the Architecture Committee has agreed on using the "Dismiss stale pull request approvals when new commits are pushed" configuration from GitHub.  It basically means that if your PR has been rebased or updated, the approvals given will be erased.

In order to minimize traumas due to the new approach, please, consider adding a note on the changes done before the rebase / force-push, and also pinging the reviewers for a subsequent round of reviews.

Thanks for your understanding!

Related issue: https://github.com/kata-containers/community/issues/249

File: src/runtime/virtcontainers/hypervisor_amd64.go
Comment (larrydewey) [inline src/runtime/virtcontainers/hypervisor_amd64.go:24]: We should probably normalize the string here. I have seen different case behavior among different distributions of Linux.

**Comment by jodh-intel** [comment]
Any update on this @jimcadden?

**Comment by jodh-intel** [comment]
@jimcadden - there are outstanding questions on this PR and concerns around vendoring. Do you plan to update or shall we close this, or mark it as wip?

**Comment by jimcadden** [comment]
> @jimcadden - there are outstanding questions on this PR and concerns around vendoring. Do you plan to update or shall we close this, or mark it as wip?

Hi @jodh-intel, @fitzthum and I are re-engineering this patch to removed the dependency concerns and to bring the protocol logic into the shim. We will close this PR and open a new one when the new patch is complete.

**Comment by jodh-intel** [comment]
Thanks for the update @jimcadden.

**Pull request closed without merging**
